src/datasources/email/email.datasource.ts

```diff
@@ -0,0 +1,157 @@
+import { EmailVerificationCode } from '@/domain/email/entities/email-verification-code.entity';
+import { Email } from '@/datasources/email/entities/email.entity';
+import { VerificationStatus } from '@/datasources/email/entities/verification.status';
+import { EmailAddressDoesNotExistError } from '@/datasources/email/errors/email-address-does-not-exist.error';
+import { EmailAlreadyVerifiedError } from '@/datasources/email/errors/email-already-verified.error';
+import { InvalidVerificationCodeError } from '@/datasources/email/errors/invalid-verification-code.error';
+import { ExpiredVerificationCodeError } from '@/datasources/email/errors/expired-verification-code.error';
+import { IConfigurationService } from '@/config/configuration.service.interface';
+
+@Injectable()
+export class EmailDatasource implements IEmailDataSource {
+  private readonly verificationExpireTimeInSeconds: number;
+
+  constructor(
+    @Inject('DB_INSTANCE') private readonly sql: postgres.Sql,
+    @Inject(IConfigurationService) configurationService: IConfigurationService,
+  ) {
+    this.verificationExpireTimeInSeconds = configurationService.getOrThrow(
+      'email.verificationExpireTimeInSeconds',
+    );
+  }
+
+  private _resetVerificationCode(
+    sql: postgres.Sql,
+    emailId: number,
+  ): postgres.PendingQuery<VerificationStatus[]> {
+    // TODO proper code generation
+    const code = Math.floor(Math.random() * 100_000);
+    return sql<VerificationStatus[]>`
+        INSERT INTO emails.verification (id, verification_code)
+        VALUES (${emailId}, ${code})
+        ON CONFLICT (id) DO UPDATE SET verification_code = ${code},
+                                       sent_on           = now()
+        RETURNING *
+    `;
+  }
+
+  private getEmail(args: {
+    chainId: string;
+    safeAddress: string;
+    signer: string;
+  }): postgres.PendingQuery<Email[]> {
+    return this.sql<Email[]>`SELECT *
+                             FROM emails.signer_emails
+                             WHERE chain_id = ${args.chainId}
+                               and safe_address = ${args.safeAddress}
+                               and signer = ${args.signer}`;
+  }
+
+  async saveEmail(args: {
+    chainId: string;
+    safeAddress: string;
+    emailAddress: string;
+    signer: string;
+  }): Promise<EmailVerificationCode> {
+    return await this.sql.begin(async (sql) => {
+      const [email] = await sql<Email[]>`
+          INSERT INTO emails.signer_emails (chain_id, email_address, safe_address, signer)
+          VALUES (${args.chainId}, ${args.emailAddress}, ${args.safeAddress}, ${args.signer})
+          RETURNING *
+      `;
+
+      const [verificationStatus] = await this._resetVerificationCode(
+        sql,
+        email.id,
+      );
+
+      return <EmailVerificationCode>{
+        emailAddress: email.email_address,
+        verificationCode: verificationStatus.verification_code,
+      };
+    });
+  }
```

Comment:
I assume that we will later implement the signature checks for these methods?

Reply:
Yes but at which level is still up for discussion. The functionality in the Email Datasource can be provided without these checks and we assume that the component will be used by one which will validate the signatures. This would decouple the database operations with the authentication part.